carbon-design-system/carbon
# feat(popover): use floating-ui for autoAlign

Closes #14493 

This PR enhances Popover's experimental `autoAlign` functionality to automatically position the floating element to be within view, even on scroll. 

This uses [`floating-ui`](https://floating-ui.com/) and relies solely on the `fixed` strategy, using `position: fixed` on the floating element. In the majority of cases, Popover floating elements will no longer be clipped by parents with `overflow: hidden`.

This implementation requires two important changes I'd like to highlight:
1. The caret must be a child of the popover content instead of a sibling, so there is a dom change here.
2. The `.#{$prefix}--popover-container` must not have `position: relative`

#### Changelog

**Changed**

- Add new alignment options to match the floating-ui api with backwards compatibility. These also more properly reflect the "logical property" naming convention of start/end instead of left/right, for instance.
- Use floating-ui for positioning the caret and floating element, apply styles via CSSOM to comply with CSP.
- Feed floating-ui the values of popover's custom properties (`offset`, caret size, etc) when possible.
- Update Popover styling so no positioning styling is used when `autoAlign={true}`
- Update ToggletipButton to use forwardRef
- Update Slug styling to be compatible with `autoAlign`

**Removed**

- There was an unused ref on Tooltip, I removed it.

#### Testing / Reviewing

- View the popover stories, there are a few autoalign experiment stories in there. Ensure it works as intended and there's no regressions.
- View the Tooltip autoalign story and ensure the autoAlign functionality works with no regressions
- View the ToggleTip autoalign story and ensure the autoAlign functionality works with no regressions
- View the slug stories and ensure ensure the autoAlign functionality works with no regressions
